packages/trader/src/App/Components/Elements/VideoPlayer/video-player.tsx

```diff
@@ -0,0 +1,281 @@
+        (!user_browser.isSafari() || (is_mobile && mobileOSDetect() !== 'iOS' && mobileOSDetect() !== 'unknown')) ??
+        true;
+
+    const [is_animated, setIsAnimated] = React.useState(true);
+    const [is_playing, setIsPlaying] = React.useState(false);
+    const [is_ended, setIsEnded] = React.useState(false);
+    const [is_muted, setIsMuted] = React.useState(false);
+    const [has_enlarged_dot, setHasEnlargedDot] = React.useState(false);
+    const [current_time, setCurrentTime] = React.useState<number>();
+    const [video_duration, setVideoDuration] = React.useState<number>();
+    const [shift_X, setShiftX] = React.useState<number>(0);
+    const [show_controls, setShowControls] = React.useState(!should_autoplay);
+
+    const video_ref = React.useRef<StreamPlayerApi>();
+    const progress_bar_filled_ref = React.useRef<HTMLDivElement>(null);
+    const progress_bar_ref = React.useRef<HTMLDivElement>(null);
+    const progress_dot_ref = React.useRef<HTMLSpanElement>(null);
+    const animation_ref = React.useRef<number>(0);
+    const timer_ref = React.useRef<ReturnType<typeof setTimeout>>();
+    const is_dragging = React.useRef<boolean>(false);
+
+    const onVolumeChange = (new_value: number) => {
+        if (!video_ref.current) return;
+
+        video_ref.current.volume = new_value;
+    };
+
+    const onPlaybackRateChange = (new_value: number) => {
+        if (!video_ref.current) return;
+
+        video_ref.current.playbackRate = new_value;
+    };
+
+    const togglePlay = (
+        e:
+            | React.MouseEvent<HTMLDivElement | HTMLButtonElement>
+            | React.KeyboardEvent<HTMLDivElement | HTMLButtonElement>
+    ) => {
+        e.stopPropagation();
+
+        if (!video_ref.current || !progress_bar_filled_ref.current) return;
+
+        if (is_playing) {
+            video_ref.current.pause();
+            setIsAnimated(false);
+            cancelAnimationFrame(animation_ref.current);
+        } else {
+            video_ref.current.play();
+            setIsAnimated(true);
+        }
+
+        setIsPlaying(prev => !prev);
+        setIsEnded(false);
+    };
+
+    const calculateNewWidth = (
+        e:
+            | React.MouseEvent<HTMLDivElement | HTMLSpanElement>
+            | React.TouchEvent<HTMLDivElement | HTMLSpanElement>
+            | TouchEvent
+            | MouseEvent
+    ) => {
+        const progress_bar = progress_bar_ref.current;
+        const client_X =
+            e.type === 'mousemove' || e.type === 'click'
+                ? (e as MouseEvent).clientX
+                : (e as TouchEvent).changedTouches[0].clientX;
+
+        let new_width =
+            ((client_X - shift_X - (progress_bar?.getBoundingClientRect().left ?? 0)) /
+                (progress_bar?.getBoundingClientRect().width ?? 0)) *
+            100;
+        if (new_width >= 100) {
+            new_width = 100;
+        }
+        if (new_width <= 0) {
+            new_width = 0;
+        }
```

Comment:
```suggestion
        if (new_width >= 100) new_width = 100;
        if (new_width <= 0) new_width = 0;
```

Reply:
Will apply here and in packages/trader/src/App/Components/Elements/VideoPlayer/volume-control.tsx